Yearn crvTricrypto trigger

mds1 opened the pull request:
Implements a trigger for the Yearn crvTricrypto vault that triggers if any one of 3 conditions occur:
1. The Yearn price per share decreases by more than 50%
2. The Curve Tricrypto virtual price decreases by more than 50%
3. The internal token balances tracked by Curve Tricrypto are more than 50% below the true token balances

All tests are run against actual mainnet contracts and overwrite storage slots to ensure tests and trigger behave as expected in a mainnet environment

mds1 committed:
First pass at Yearn crvTricrypto trigger

mds1 committed:
Add balances check for tricrypto trigger

mds1 committed:
Setup tricrypto tests to use mainnet contracts

mds1 committed:
Get pricePerShare and virtual_price tests passing

mds1 committed:
Add balance check tests

mds1 committed:
Add deploy script + update readme with new trigger

apbendi commented on contracts/YearnCrvTricrypto.sol:
```diff
@@ -0,0 +1,120 @@
+   * @return True if balances are out of tolerance and trigger should be toggled
+   */
+  function checkCurveBalances() internal view returns (bool) {
+    if (usdt.balanceOf(address(curve)) < ((curve.balances(usdtIndex) * virtualPriceTol) / scale)) return true;
```
Why if statements instead of:

```solidity
return 
       (usdt.balanceOf(address(curve)) < ((curve.balances(usdtIndex) * virtualPriceTol) / scale)) || 
       (wbtc.balanceOf(address(curve)) < ((curve.balances(wbtcIndex) * virtualPriceTol) / scale)) ||
       (weth.balanceOf(address(curve)) < ((curve.balances(wethIndex) * virtualPriceTol) / scale));
```

mds1 replied:
No reason really, just defaulted to it 🤷

Your way probably saves a tiny amount of gas too, so switched to that in 143b76896f7fd7aca308c7553949b7422e18af5d

apbendi commented on contracts/YearnCrvTricrypto.sol:
```diff
@@ -0,0 +1,120 @@
+
+  // --- Tolerances ---
+  /// @dev Scale used to define percentages. Percentages are defined as tolerance / scale
+  uint256 public constant scale = 1000;
```
Just curious how you chose 10e3 as the scale factor for defining these percentages? Just kind of a "good enough" or was there a calculation involved to ensure rounding errors would necessarily fall below the threshold concerned at that scale?

mds1 replied:
It's pretty arbitrary. The other Yearn trigger we used 1e18, but multiplying by e.g. 5e17 to divide by 1e18 felt unnecessary and more likely to result in overflow. So for this one I used 1000 to reduce overflow but still give enough precision to check e.g. a 50.1% decrease if you wanted

apbendi reviewed (commented):
Overall looks good! These triggers are getting more sophisticated ;)

In addition to a couple small comment, I'm getting the following error when running the test suite:

```
  1) YearnCrvTricrypto
       checkAndToggleTrigger
         properly updates the saved state:

      AssertionError: expected '27299456984729344564195' to equal '27299455696392394563121'
      + expected - actual

      -27299456984729344564195
      +27299455696392394563121

      at Context.<anonymous> (test/YearnCrvTricrypto.test.ts:196:50)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at runNextTicks (internal/process/task_queues.js:62:3)
      at listOnTimeout (internal/timers.js:523:9)
      at processTimers (internal/timers.js:497:7)
```

mds1 commented:
> In addition to a couple small comment, I'm getting the following error when running the test suite:

Interesting. I see this now too. This wasn't happening before so it must be an artifact of the fact that the mainnet state is different now than it was earlier today. I can't tell if this is one argument for or against pinning tests to a block instead of always using the most recent block 🤔

The difference in values is large enough that I don't want to simply chalk it up to rounding error just yet, so I'll take a closer look at this tomorrow. For reference, current block number is 12867033 in case we need to reproduce on-demand

mds1 commented:
Re-running the tests right now at block 12870054 and all tests pass. Setting the Hardhat config to run tests against block 12867033 and that test fails again. Have not yet looked into exactly why it's failing though

mds1 committed:
Simplify return condition when checking curve balances

mds1 committed:
Update trigger description in deploy script

mds1 commented:
Here's what seems to be happening. The original test code is below:

```typescript
// Update values
await setYearnTotalSupply(10n ** 18n); // don't set them too high or trigger will toggle
const newPricePerShare = await yCrvTricrypto.pricePerShare();
await setCrvTotalSupply(10n ** 18n);
const newVirtualPrice = await crvTricrypto.get_virtual_price();
      
// Call checkAndToggleTrigger to simulate someone using the protocol
await trigger.checkAndToggleTrigger(); // PROBLEM: reading pricePerShare before and after this call differs
expect(await trigger.isTriggered()).to.be.false; // sanity check

// Verify the new state
const currentPricePerShare = await trigger.lastPricePerShare();
expect(currentPricePerShare.toString()).to.equal(newPricePerShare.toString()); // bigint checks are flaky with chai
const currentVirtualPrice = await trigger.lastVirtualPrice();
expect(currentVirtualPrice.toString()).to.equal(newVirtualPrice.toString());
```

You'll see the comment here that reading pricePerShare before and after our call to `checkAndToggle` differs.

We know from above that running the tests at block 12870054 passes, and at block 12867033 fails. Let's use [seth](https://github.com/dapphub/dapptools/tree/master/src/seth) to read the price per share around each of these block numbers to see what's going on:

```sh
yearn=0x3D980E50508CFd41a13837A60149927a11c03731
method="pricePerShare()(uint256)" # view function with no inputs and returns a uint256
goodBlock=12870054
badBlock=12867033

# Check good block first
seth call $yearn $method --block "$(($goodBlock-1))"  # returns 1015310673694155268
seth call $yearn $method --block "$(($goodBlock))"    # returns 1015310673694155268
seth call $yearn $method --block "$(($goodBlock+1))"  # returns 1015310673694155268

# Check bad block first
seth call $yearn $method --block "$(($badBlock-1))"  # returns 1015102738066755748
seth call $yearn $method --block "$(($badBlock))"    # returns 1015102869802727549
seth call $yearn $method --block "$(($badBlock+1))"  # returns 1015103468602599367
```

Notice how the price per share is constant around the block number when the test passed, and variable around the block number when it failed. And note that in these tests we overwrite the total supply value in storage, and then price per share is a computed getter based on total supply and various other parameters of the vault. 

Hardhat increments the block number after each transaction, so the problem was that we:
- Read price per share from block X
- Call checkAndToggleTrigger, which mines a block and stores price per share from block X+1
- If Yearn vault parameters differ between block X and X+1 on mainnet, then the `pricePerShare` is will differ between blocks X and X+1

The simplest solution is to just read the expected price after toggling the trigger—since this test is comparing what's read from the Yearn vault to what's stored in the trigger, this is ok. I pushed this fix in the most recent commit. Another option would be to turn off Hardhat's automine so we have more control over block number, but that's more complex and not needed here

mds1 committed:
Fix flaky test dependent on mainnet state

Details: https://github.com/ScopeLift/cozy-triggers/pull/2\#issuecomment-884246561

mds1 committed:
Add JumpRateModelV2 interest rate model from cozy-protocol repo

mds1 committed:
Update Yearn CrvTricrypto deploy script to use new IR model

mds1 committed:
Update IR model params to change borrow rate at kink

apbendi reviewed (approved):
Tests pass, changes look good. Thanks 👍
